test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java

```diff
@@ -52,4 +59,161 @@ public void writeTo(StreamOutput out) throws IOException {
+                {
+                    otherBuilder.field("field1", "value1");
+                    otherBuilder.field("field2", "differentValue2");
+                }
+                otherBuilder.endObject();
+            }
+            otherBuilder.endObject();
+            AssertionError error = expectThrows(AssertionError.class,
+                    () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
+            assertEquals("different xContent at [foo/field2] expected:<[v]alue2> but was:<[differentV]alue2>", error.getMessage());
+        }
+        {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            {
+                builder.startArray("foo");
+                {
+                    builder.value("one");
+                    builder.value("two");
+                    builder.value("three");
+                }
+                builder.endArray();
+            }
+            builder.endObject();
+
+            XContentBuilder otherBuilder = JsonXContent.contentBuilder();
+            otherBuilder.startObject();
+            {
+                otherBuilder.startArray("foo");
+                {
+                    otherBuilder.value("one");
+                    otherBuilder.value("two");
+                    otherBuilder.value("four");
+                }
+                otherBuilder.endArray();
+            }
+            otherBuilder.endObject();
+            AssertionError error = expectThrows(AssertionError.class,
+                    () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
+            assertEquals("different xContent at [foo/2] expected:<[three]> but was:<[four]>", error.getMessage());
+        }
+        {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            {
+                builder.startArray("foo");
+                {
+                    builder.value("one");
+                    builder.value("two");
+                    builder.value("three");
+                }
+                builder.endArray();
+            }
+            builder.endObject();
+
+            XContentBuilder otherBuilder = JsonXContent.contentBuilder();
+            otherBuilder.startObject();
+            {
+                otherBuilder.startArray("foo");
+                {
+                    otherBuilder.value("one");
+                    otherBuilder.value("two");
+                }
+                otherBuilder.endArray();
+            }
+            otherBuilder.endObject();
+            AssertionError error = expectThrows(AssertionError.class,
+                    () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
+            assertEquals("different list values at [foo]: [one, two, three] vs. [one, two] expected:<3> but was:<2>", error.getMessage());
```

Comment:
here too, maybe go straight to what was expected and what was found. I don't understand which of the two is what in this error message, and the last part confuses me. Maybe it's me.